src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs

```diff
@@ -377,20 +386,28 @@ public override Matcher Build()
             if (node.PolicyEdges != null && node.PolicyEdges.Count > 0)
             {
                 var policyBuilder = new PolicyJumpTableBuilder(node.NodeBuilder);
-                tableBuilders[stateIndex] = (pathBuilder, policyBuilder);
+                tableBuilders[currentStateIndex] = (pathBuilder, policyBuilder);
 
                 foreach (var kvp in node.PolicyEdges)
                 {
                     policyBuilder.AddEntry(kvp.Key, Transition(kvp.Value));
                 }
             }
 
-            return stateIndex;
+            return currentStateIndex;
 
             int Transition(DfaNode next)
             {
                 // Break cycles
-                return ReferenceEquals(node, next) ? stateIndex : AddNode(next, states, tableBuilders);
+                if (ReferenceEquals(node, next))
+                {
+                    return _stateIndex;
+                }
+                else
+                {
+                    _stateIndex++;
+                    return AddNode(next, states, tableBuilders);
```

Comment:
AddNode should take the index as a parameter rather than tunneling through a field.

Reply:
That is what I originally had but it becomes unwieldy because it has to be passed by ref.

See: https://github.com/aspnet/Routing/pull/767/commits/082dafae94a790453059fb033e86177faa4d89d4

Are you sure?

Reply:
ah I see, yeah its fine as is